src/getThreadList.js

```diff
@@ -1,73 +1,211 @@
-      };
-    }
+// "FF8C0077" -> "8C0077"
+function formatColor(color) {
+  if (color && color.match(/^(?:[0-9a-fA-F]{8})$/g)) {
+    return color.slice(2);
+  }
+  return color;
+}
 
-    if (end <= start) end = start + 20;
+function getThreadName(t) {
+  if (t.name || t.thread_key.thread_fbid) return t.name;
 
-    var form = {
-      client: "mercury"
-    };
+  for (let p of t.all_participants.nodes) {
+    if (p.messaging_actor.id === t.thread_key.other_user_id) return p.messaging_actor.name;
+  }
+}
 
-    form[type + "[offset]"] = start;
-    form[type + "[limit]"] = end - start;
+function mapNicknames(customizationInfo) {
+  return customizationInfo?customizationInfo.participant_customizations.map(u => {
+    return {"userID": u.participant_id, "nickname": u.nickname};
+  }):[];
+}
 
-    if (ctx.globalOptions.pageID) {
-      form.request_user_id = ctx.globalOptions.pageID;
+function formatThreadList(data) {
+  return data.map(t => {
+    return {
+      threadID: utils.formatID(t.thread_key.thread_fbid || t.thread_key.other_user_id),
+      name: getThreadName(t),
+      unreadCount: t.unread_count,
+      messageCount: t.messages_count,
+      imageSrc: t.image?t.image.uri:null,
+      emoji: t.customization_info?t.customization_info.emoji:null,
+      color: formatColor(t.customization_info?t.customization_info.outgoing_bubble_color:null),
+      nicknames: mapNicknames(t.customization_info),
+      muteUntil: t.mute_until,
+      participants: formatParticipants(t.all_participants),
+      participantIDs: formatParticipants(t.all_participants).map(participant => participant.userID),
+      adminIDs: t.thread_admins.map(a => a.id),
+      folder: t.folder,
+      isGroup: t.thread_type === "GROUP",
+      // rtc_call_data: t.rtc_call_data, // TODO: format and document this
+      // isPinProtected: t.is_pin_protected, // feature from future? always false (2018-04-04)
+      customizationEnabled: t.customization_enabled, // false for ONE_TO_ONE with Page or ReducedMessagingActor
+      participantAddMode: t.participant_add_mode_as_string, // "ADD" if "GROUP" and null if "ONE_TO_ONE"
+      montageThread: t.montage_thread?Buffer.from(t.montage_thread.id,"base64").toString():null, // base64 encoded string "message_thread:0000000000000000"
+      // it is not userID nor any other ID known to me...
+      // can somebody inspect it? where is it used?
+      // probably Messenger Day uses it
+      reactionsMuteMode: t.reactions_mute_mode,
+      mentionsMuteMode: t.mentions_mute_mode,
+      isArchived: t.has_viewer_archived,
+      isSubscribed: t.is_viewer_subscribed,
+      timestamp: t.updated_time_precise, // in miliseconds
+      // isCanonicalUser: t.is_canonical_neo_user, // is it always false?
+      // TODO: how about putting snippet in another object? current implementation does not handle every possibile message type etc.
+      snippet: t.last_message.nodes[0].snippet,
```

Comment:
Can we make this a bit more robust and check the presence of these elements before accessing? (same goes for everything below)

Reply:
For everything? :/
Why not only for these with nodes[0]?

Reply:
done